### PR TITLE
Build Strategy

### PR DESCRIPTION
**Pull request by ghost**
This PR separates the BVH construction algorithm from the actual BVH. This will allow there to be multiple BVH construction algorithms without putting them all in the BVH.h file. The algorithms for building the BVH are referred to as "strategies", based on the strategy design pattern. There are currently two:

### Build Strategy 0

This is sort of a "dummy" build strategy. All it does is build a single bounding box around the primitives. It's main purpose will probably be to determine the effectiveness of other strategies by relative comparison.

### Build Strategy 1

This is the original BVH construction algorithm that was moved out of `BVH.h`. I wasn't quite sure if the algorithm had a name, but luckily the algorithms aren't differentiated by class name and they differentiated by a "variant index" template parameter. For example:

```cxx
/* Refers to the "dummy" build strategy */
BuildStrategy<float, 0> strategy0;

/* Refers to the original build strategy */
BuildStrategy<float, 1> strategy1;

/* This will be the build algorithm referred to in issue #14. */
BuildStrategy<float, 2> strategy2;
```

There's a type definition for using this build algorithm, called `DefaultBuilder`.

### Changes to BVH

The BVH structure went through some changes as well. Admittedly, this probably could have been left out to keep the PR a little bit leaner. Here's the gist of it:

 - BVH no longer stores the primitives in a `std::vector` instance. I'd like to gradually remove the STL dependency, so that this project can be built with CUDA toolkit.
 - Some extraneous data, such as the number of leafs in the BVH, has been removed. You can still get the number of leafs in the BVH by calling `BVH::countLeafs()`.

### Example

Here's a code example that uses the new class.

```cxx
std::vector<Sphere> spheres;

/* Add the spheres */

SphereBoxConverter box_converter;

FastBVH::DefaultBuilder builder;

auto bvh = builder(spheres, box_converter);

/* traverse bvh */
```

This addresses the first item in #14.

### CODE REVIEW

File: include/FastBVH/BuildStrategy1.h [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+
+  using namespace Strategy1;
+
+  // The threshold hold at which a leaf is made in the BVH.
+  static constexpr uint32_t leaf_size = 4;
+
+  BuildStack todo;
+
+  const uint32_t Untouched    = 0xffffffff;
+  const uint32_t TouchedTwice = 0xfffffffd;
```
Comment (brandonpelfrey) [inline include/FastBVH/BuildStrategy1.h]: I'm going to create an issue to better document how this works.

File: tests/BuildStrategy.cxx [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+#include <list>
+
+using namespace FastBVH;
+
+namespace {
+
+//! \brief This is just boilerplate code for google test
+//! to be able to run template tests.
+template <typename T>
+class BuildStrategyTest : public ::testing::Test {
```
Comment (brandonpelfrey) [inline tests/BuildStrategy.cxx:15]: Awesome :+1: Tests will be great. We can get this added to the GitHub Actions to run these checks on new PRs.

File: include/FastBVH/BuildStrategy1.h (hunk 1 above)
Comment (ghost) [reply include/FastBVH/BuildStrategy1.h]: Thanks! I was wondering about that

File: tests/BuildStrategy.cxx (hunk 2 above)
Comment (ghost) [reply tests/BuildStrategy.cxx:15]: Awesome! I didn't write any tests yet for the existing code, I figured if it works, it works. I'll be proactive writing tests for new code though.

**Comment by ghost** [comment]
Do you have a preference between single commit PR and many-commit PRs? I can squash these once I'm done handling the merge conflicts.

**Review by brandonpelfrey — COMMENTED** [review]
Looks great! Only general comment will be for us to format from the `.clang-format` and the general line spacing guidance for readability. All of this looks like a huge leap in the right direction. After this I'll create an issue to better track what I'm working on so we can avoid merge problems.

**Comment by ghost** [comment]
I'll be sure to run `clang-format` and make the new code consistent with the new line break style.

**Comment by brandonpelfrey** [comment]
> Do you have a preference between single commit PR and many-commit PRs? I can squash these once I'm done handling the merge conflicts.

Good to squash in my mind, at least into logical blocks of functionality if possible. It doesn't need to be single commit, but each commit should be meaningful.

**Comment by ghost** [comment]
@brandonpelfrey I kept things simple and just squashed all my commits, then made a merge commit. In the future, I'll use fewer commits without squashing so that it's more organized.

**Review by brandonpelfrey — APPROVED** [review]
Nice set of changes. Merging 👍

**Pull request merged**